returnn/tf/layers/basic.py

```diff
@@ -4769,39 +4773,57 @@ def __init__(self, eval, **kwargs):
 
 class CompareLayer(LayerBase):
   """
-  This layer performs an elementwise comparison of all provided sources, and returnns a boolean vector
-  of the same shape.
-  If the `value` parameter is provided, all elements in the provided sources are also compared to this value.
+  Compares element-wise the tokens of all input sequences among themselves and/or with a specified given value.
+  The comparisons are performed in a chain according to the order in which they are listed. 
 
-  If more than two sources are provided, each additional source will be compared against the first source,
-  and the results will be combined by a logical conjunction.
+  Example:
+    {"class": "compare", "from": ["i1", "i2"], "value": val, "kind": "less"}
+    computes i1 < i2 < val and it is true only if the whole chain of operations is true.
+    The final result is the logical "and" of all comparisons. Note that `value` is the last element to be compared to.
+  
+  A common example usage is the `end` layer in a rec subnetwork to specify the stopping criterion,
+  e.g. the last generated token is equal to the end-of-sentence token:
+
+    "output": {"class": "rec", "from": [], "unit": {
+        .
+        .
+        .
+        "end": {"class": "compare", "from": "output", "value": end_of_sentence_id}
+    }, "target": "classes0"},
   """
   layer_class = "compare"
 
   def __init__(self, kind="equal", value=None, **kwargs):
     """
-    :param str kind: Which comparison operation to use, e.g. "equal" or other supported tf comparison ops
+    :param str kind: Which comparison operation to use, e.g. "equal", "greater", "less"
+    or other supported tf comparison ops
     :param float|int|None value: if specified, will also compare to this
     """
     super(CompareLayer, self).__init__(**kwargs)
     assert len(self.sources) >= 1
     if value is None:
-      assert len(self.sources) >= 2
+      assert len(self.sources) >= 2, "{} requires at least two elements to compare.".format(self.__name__)
     op = getattr(tf, kind)  # e.g. tf.equal
     from returnn.tf.util.basic import swapaxes
-    x = self.sources[0].output.placeholder
+    common_data = Data.get_common_data([s.output for s in self.sources])
```

Comment:
Note that when you do that here, you should also fix `get_out_data_from_opts`. See `CombineLayer.get_out_data_from_opts`, which you can basically copy here.